File: ai-agent-service/services/workflow_executor.py
"""
Workflow execution engine for backend nodes
Handles AI Agent, API Call, and Handoff node execution
"""

from typing import Dict, List, Optional, Any
from dataclasses import dataclass
import httpx


@dataclass
class WorkflowState:
    """Current workflow execution state"""
    current_node_id: str
    variables: Dict[str, Any]
    execution_history: List[str]
    user_input: Optional[str] = None


@dataclass
class NodeExecutionResult:
    """Result of executing a single node"""
    messages: List[str]
    next_node_id: Optional[str]
    variables: Optional[Dict[str, Any]] = None
    requires_input: bool = False
    is_complete: bool = False


class WorkflowExecutor:
    """
    Backend workflow executor for hybrid execution
    Executes backend-required nodes (AI Agent, API Call, Handoff)
    """
    
    def __init__(
        self,
        workflow_definition: Dict[str, Any],
        session_id: str,
        project_id: str,
        llm_service=None,
        kb_service=None
    ):
        self.workflow = workflow_definition
        self.session_id = session_id
        self.project_id = project_id
        self.llm_service = llm_service
        self.kb_service = kb_service
        self.state: Optional[WorkflowState] = None
    
    def load_state(self, state_dict: Dict[str, Any]):
        """Load execution state from frontend"""
        self.state = WorkflowState(
            current_node_id=state_dict.get('currentNodeId', ''),
            variables=state_dict.get('variables', {}),
            execution_history=state_dict.get('executionHistory', []),
            user_input=state_dict.get('userInput')
        )
    
    async def execute_current_node(self) -> NodeExecutionResult:
        """Execute the current node"""
        if not self.state:
            raise ValueError("State not loaded")
        
        # Find current node
        current_node = self._find_node(self.state.current_node_id)
        if not current_node:
            return NodeExecutionResult(
                messages=["Error: Invalid workflow state"],
                next_node_id=None,
                is_complete=True
            )
        
        # Execute based on node type
        node_type = current_node.get('type')
        
        if node_type == 'ai-agent':
            return await self._execute_ai_agent_node(current_node)
        elif node_type == 'api-call':
            return await self._execute_api_call_node(current_node)
        elif node_type == 'handoff':
            return await self._execute_handoff_node(current_node)
        else:
            # Shouldn't reach here in hybrid execution
            return NodeExecutionResult(
                messages=[],
                next_node_id=self._get_next_node_id(current_node),
                is_complete=False
            )
    
    async def process_user_input(self, user_input: str) -> NodeExecutionResult:
        """Process user input and execute workflow"""
        # Initialize state if not exists
        if not self.state:
            start_node = next((n for n in self.workflow['nodes'] if n['type'] == 'start'), None)
            if not start_node:
                return NodeExecutionResult(
                    messages=["Error: No start node found"],
                    next_node_id=None,
                    is_complete=True
                )
            
            self.state = WorkflowState(
                current_node_id=start_node['id'],
                variables=self.workflow.get('variables', {}),
                execution_history=[start_node['id']],
                user_input=user_input
            )
            
            # Move to next node after start
            next_id = self._get_next_node_id(start_node)
            if next_id:
                self.state.current_node_id = next_id
        else:
            self.state.user_input = user_input
        
        return await self.execute_current_node()
    
    async def _execute_ai_agent_node(self, node: Dict[str, Any]) -> NodeExecutionResult:
        """Execute AI Agent node"""
        try:
            # Get node configuration
            use_kb = node.get('data', {}).get('useKnowledgeBase', True)
            custom_prompt = node.get('data', {}).get('prompt', '')
            
            # Get context from knowledge base if enabled
            context = ""
            if use_kb and self.kb_service:
                context = await self.kb_service.get_relevant_context(
                    self.state.user_input or '',
                    self.project_id
                )
            
            # Generate response using LLM
            if self.llm_service:
                # Build prompt
                prompt = custom_prompt if custom_prompt else self.state.user_input or ''
                
                response = await self.llm_service.generate_response(
                    query=prompt,
                    context=context,
                    history=[],  # Could load from session
                    persona={}   # Could load from project config
                )
                
                messages = [response]
            else:
                messages = ["AI Agent: " + (self.state.user_input or "No input")]
            
            # Get next node
            next_node_id = self._get_next_node_id(node)
            
            return NodeExecutionResult(
                messages=messages,
                next_node_id=next_node_id,
                variables=self.state.variables,
                is_complete=not next_node_id
            )
            
        except Exception as e:
            print(f"Error executing AI Agent node: {e}")
            return NodeExecutionResult(
                messages=[f"Error: {str(e)}"],
                next_node_id=None,
                is_complete=True
            )
    
    async def _execute_api_call_node(self, node: Dict[str, Any]) -> NodeExecutionResult:
        """Execute API Call node"""
        try:
            # Get node configuration
            method = node.get('data', {}).get('method', 'GET')
            url = node.get('data', {}).get('url', '')
            headers = node.get('data', {}).get('headers', {})
            body = node.get('data', {}).get('body', '')
            response_var = node.get('data', {}).get('responseVariable', 'api_response')
            
            if not url:
                return NodeExecutionResult(
                    messages=["Error: No URL specified for API call"],
                    next_node_id=None,
                    is_complete=True
                )
            
            # Make API call
            async with httpx.AsyncClient() as client:
                if method == 'GET':
                    response = await client.get(url, headers=headers, timeout=10.0)
                elif method == 'POST':
                    response = await client.post(url, headers=headers, content=body, timeout=10.0)
                elif method == 'PUT':
                    response = await client.put(url, headers=headers, content=body, timeout=10.0)
                elif method == 'DELETE':
                    response = await client.delete(url, headers=headers, timeout=10.0)
                else:
                    raise ValueError(f"Unsupported method: {method}")
                
                # Store response in variables
                self.state.variables[response_var] = response.text
                
                # Get next node
                next_node_id = self._get_next_node_id(node)
                
                return NodeExecutionResult(
                    messages=[],  # API calls don't send messages to user
                    next_node_id=next_node_id,
                    variables=self.state.variables,
                    is_complete=not next_node_id
                )
                
        except Exception as e:
            print(f"Error executing API Call node: {e}")
            return NodeExecutionResult(
                messages=[f"API Error: {str(e)}"],
                next_node_id=None,
                is_complete=True
            )
    
    async def _execute_handoff_node(self, node: Dict[str, Any]) -> NodeExecutionResult:
        """Execute Handoff node"""
        target = node.get('data', {}).get('target', 'human')
        message = node.get('data', {}).get('message', 'Transferring to human agent...')
        
        # In a real implementation, this would trigger handoff logic
        # For now, just return a message
        return NodeExecutionResult(
            messages=[message],
            next_node_id=None,  # Handoff ends the workflow
            is_complete=True
        )
    
    def _find_node(self, node_id: str) -> Optional[Dict[str, Any]]:
        """Find a node by ID"""
        return next((n for n in self.workflow['nodes'] if n['id'] == node_id), None)
    
    def _get_next_node_id(self, node: Dict[str, Any]) -> Optional[str]:
        """Get the next node ID from edges"""
        node_id = node['id']
        edge = next((e for e in self.workflow['edges'] if e['source'] == node_id), None)
        return edge['target'] if edge else None
    
    def get_state(self) -> Dict[str, Any]:
        """Get current state as dict"""
        if not self.state:
            return {}
        
        return {
            'currentNodeId': self.state.current_node_id,
            'variables': self.state.variables,
            'executionHistory': self.state.execution_history,
            'userInput': self.state.user_input
        }

File: chat-platform/src/lib/WorkflowExecutor.ts
import {
    WorkflowDefinition,
    WorkflowNode,
    WorkflowEdge,
    WorkflowExecutionState,
    NodeExecutionResult,
} from './workflow-types';
import { getNextNodeId } from './workflow-utils';

/**
 * Frontend workflow executor for hybrid execution
 * Executes frontend-compatible nodes and delegates to backend when needed
 */
export class WorkflowExecutor {
    private workflow: WorkflowDefinition;
    private state: WorkflowExecutionState;
    private projectId: string;
    private sessionId: string;

    constructor(
        workflow: WorkflowDefinition,
        projectId: string,
        sessionId: string = 'default'
    ) {
        this.workflow = workflow;
        this.projectId = projectId;
        this.sessionId = sessionId;

        // Initialize state with start node
        const startNode = workflow.nodes.find(n => n.type === 'start');
        this.state = {
            currentNodeId: startNode?.id || '',
            variables: { ...workflow.variables },
            executionHistory: [],
            userInput: undefined,
        };
    }

    /**
     * Start workflow execution
     */
    async start(): Promise<NodeExecutionResult> {
        const startNode = this.workflow.nodes.find(n => n.type === 'start');
        if (!startNode) {
            throw new Error('Workflow must have a start node');
        }

        this.state.currentNodeId = startNode.id;
        this.state.executionHistory.push(startNode.id);

        // Move to next node after start
        const nextNodeId = getNextNodeId(startNode, this.workflow.edges);
        if (!nextNodeId) {
            return {
                messages: [],
                nextNodeId: null,
                isComplete: true,
            };
        }

        this.state.currentNodeId = nextNodeId;
        return this.executeCurrentNode();
    }

    /**
     * Process user input and continue execution
     */
    async processUserInput(input: string): Promise<NodeExecutionResult> {
        this.state.userInput = input;
        return this.executeCurrentNode();
    }

    /**
     * Execute the current node
     */
    private async executeCurrentNode(): Promise<NodeExecutionResult> {
        const currentNode = this.workflow.nodes.find(n => n.id === this.state.currentNodeId);
        if (!currentNode) {
            return {
                messages: ['Error: Invalid workflow state'],
                nextNodeId: null,
                isComplete: true,
            };
        }

        this.state.executionHistory.push(currentNode.id);

        // Check if node requires backend execution
        if (this.requiresBackendExecution(currentNode)) {
            return this.executeOnBackend(currentNode);
        }

        // Execute on frontend
        return this.executeNodeFrontend(currentNode);
    }

    /**
     * Check if node requires backend execution
     */
    private requiresBackendExecution(node: WorkflowNode): boolean {
        return ['ai-agent', 'api-call', 'handoff'].includes(node.type);
    }

    /**
     * Execute node on backend (hybrid execution)
     */
    private async executeOnBackend(node: WorkflowNode): Promise<NodeExecutionResult> {
        try {
            const response = await fetch('/api/chat', {
                method: 'POST',
                headers: { 'Content-Type': 'application/json' },
                body: JSON.stringify({
                    query: this.state.userInput || '',
                    project_id: this.projectId,
                    session_id: this.sessionId,
                    workflow_state: {
                        currentNodeId: node.id,
                        variables: this.state.variables,
                        executionHistory: this.state.executionHistory,
                    },
                }),
            });

            const data = await response.json();

            // Update state with backend result
            if (data.node_result) {
                const result = data.node_result;
                this.state.variables = { ...this.state.variables, ...result.variables };
                this.state.currentNodeId = result.nextNodeId || '';

                return {
                    messages: result.messages || [data.response],
                    nextNodeId: result.nextNodeId,
                    variables: this.state.variables,
                    isComplete: !result.nextNodeId,
                };
            }

            return {
                messages: [data.response || 'No response'],
                nextNodeId: null,
                isComplete: true,
            };
        } catch (error) {
            console.error('Backend execution error:', error);
            return {
                messages: ['Error executing workflow'],
                nextNodeId: null,
                isComplete: true,
            };
        }
    }

    /**
     * Execute node on frontend
     */
    private async executeNodeFrontend(node: WorkflowNode): Promise<NodeExecutionResult> {
        switch (node.type) {
            case 'message':
                return this.executeMessageNode(node);
            case 'input':
                return this.executeInputNode(node);
            case 'condition':
                return this.executeConditionNode(node);
            case 'variable-set':
                return this.executeVariableSetNode(node);
            case 'end':
                return this.executeEndNode(node);
            default:
                const nextNodeId = getNextNodeId(node, this.workflow.edges);
                return {
                    messages: [],
                    nextNodeId,
                    isComplete: !nextNodeId,
                };
        }
    }

    /**
     * Execute message node
     */
    private executeMessageNode(node: WorkflowNode): NodeExecutionResult {
        const message = node.data.message || 'No message';
        const nextNodeId = getNextNodeId(node, this.workflow.edges);

        return {
            messages: [message],
            nextNodeId,
            isComplete: !nextNodeId,
        };
    }

    /**
     * Execute input node
     */
    private executeInputNode(node: WorkflowNode): NodeExecutionResult {
        // Input node waits for user input
        if (!this.state.userInput) {
            return {
                messages: [node.data.prompt || 'Please enter your message'],
                nextNodeId: node.id, // Stay on this node
                requiresInput: true,
                isComplete: false,
            };
        }

        // User provided input, move to next node
        const nextNodeId = getNextNodeId(node, this.workflow.edges);
        return {
            messages: [],
            nextNodeId,
            isComplete: !nextNodeId,
        };
    }

    /**
     * Execute condition node
     */
    private executeConditionNode(node: WorkflowNode): NodeExecutionResult {
        const conditions = node.data.conditions || [];
        const input = this.state.userInput || '';

        // Evaluate conditions
        for (const condition of conditions) {
            if (this.evaluateCondition(condition, input)) {
                return {
                    messages: [],
                    nextNodeId: condition.targetNodeId,
                    isComplete: !condition.targetNodeId,
                };
            }
        }

        // No condition matched, use default
        const defaultTarget = node.data.defaultTargetNodeId;
        return {
            messages: [],
            nextNodeId: defaultTarget || null,
            isComplete: !defaultTarget,
        };
    }

    /**
     * Evaluate a condition
     */
    private evaluateCondition(condition: any, input: string): boolean {
        switch (condition.type) {
            case 'keyword':
                return input.toLowerCase().includes(condition.value.toLowerCase());
            case 'regex':
                try {
                    const regex = new RegExp(condition.value, 'i');
                    return regex.test(input);
                } catch {
                    return false;
                }
            case 'variable':
                return this.state.variables[condition.value] === true;
            default:
                return false;
        }
    }

    /**
     * Execute variable set node
     */
    private executeVariableSetNode(node: WorkflowNode): NodeExecutionResult {
        const { variableName, value, valueType } = node.data;

        if (variableName) {
            if (valueType === 'static') {
                this.state.variables[variableName] = value;
            } else {
                // Expression evaluation (simple version)
                this.state.variables[variableName] = this.evaluateExpression(value);
            }
        }

        const nextNodeId = getNextNodeId(node, this.workflow.edges);
        return {
            messages: [],
            nextNodeId,
            variables: this.state.variables,
            isComplete: !nextNodeId,
        };
    }

    /**
     * Execute end node
     */
    private executeEndNode(node: WorkflowNode): NodeExecutionResult {
        return {
            messages: [],
            nextNodeId: null,
            isComplete: true,
        };
    }

    /**
     * Simple expression evaluator
     */
    private evaluateExpression(expr: string): any {
        // Very basic - just return the expression for now
        // In production, use a safe expression evaluator
        return expr;
    }

    /**
     * Get current state
     */
    getState(): WorkflowExecutionState {
        return { ...this.state };
    }
}
